openedx/features/qverse_features/registration/signals.py

```diff
@@ -1,20 +1,170 @@
+from django.db import IntegrityError, transaction
 from django.db.models.signals import post_save
 from django.dispatch import receiver
 
-from openedx.features.qverse_features.registration.models import BulkUserRegistration
+from openedx.features.qverse_features.registration.models import BulkUserRegistration, QVerseUserProfile, Department
+from student.models import UserProfile
 
-UserInfo = namedtuple('UserInfo',
-                      ['username', 'email','first_name', 'last_name',
-                       'other_name', 'mobile_number', 'level',
-                       'programme_and_department']
-                     )
+
+LOGGER = logging.getLogger(__name__)
 
 
 @receiver(post_save, sender=BulkUserRegistration)
-def make_users_from_csv_file(sender, instance, created, **kwargs):
-    # TODO: this would be implemented later once the basic architecture is final.
-    pass
+def create_users_from_csv_file(sender, instance, created, **kwargs):
+    """
+    Creates/Updates users and their profiles.
+
+    Reads data from given csv file having user details. Then creates/updates
+    edx user, edx user profile and qverse user profile accordingly. It also
+    writes the status of creation/update on the given file by adding an
+    extra column named 'status' in it. Also handles the exceptions raised by
+    the functions that it calls.
+
+    Arguments:
+        sender (ModelBase): responsible to initiate the signal
+        instance (BulkUserRegistration): newly created/updated object of BulkUserRegistration
+        created (bool): Is created/updated?
+        kwargs (dict): Other info
+
+    Returns:
+        None
+    """
+    csv_file = None
+    try:
+        csv_file = open(instance.admission_file.path, 'r')
+    except IOError as error:
+        LOGGER.exception('({}) --- {}'.format(error.filename, error.strerror))
+        return
+    dict_reader = DictReader(csv_file, delimiter=',')
+    # The header of uploaded csv might be in uppercase, lowercase or in camel case
+    # so to make our code independent to case just turn all the keys of reader into lower case.
+    # Also in python standard library for csv there is an option named skipinitialwhitespaces
+    # which is only applied for leading whitespaces and there is no option for trailing
+    # whitespaces. So, we will have to handle that case ourselves
+    reader = (dict((k.strip().lower(), v.strip()) for k, v in row.items()) for row in dict_reader)
+    output_file_rows = []
+    for row in reader:
+        try:
+            output_file_rows.append(row)
+            output_file_rows[-1].update({'status': ''})
+            # Creating/Updating edX User, edX User Profile and QVerse User Profile
+            # will be an atomic operation. If one operation fails the previous
+            # successfull operations will be reverted and no changes will be applied.
+            with transaction.atomic():
+                edx_user, _ = _create_or_update_edx_user(row)
```

Comment:
```suggestion
                edx_user = _create_or_update_edx_user(row)
```

You are not testing the code?

Reply:
done.